File: Assignments/Assignment 1/task2.py

from pyspark import SparkConf, SparkContext
import json
import sys
import time

def custom_partitioner(key):
    # Using a simple hash-based partitioning based on the business_id
    return hash(key)

def main():
    # Initialize Spark
    conf = SparkConf().setAppName("Assignment1_Task2")
    sc = SparkContext(conf=conf)
    sc.setLogLevel("ERROR")

    # Set the input and output path
    input_filepath = sys.argv[1]
    output_filepath = sys.argv[2]
    n_partition = int(sys.argv[3])

    # Read the input file
    reviews = sc.textFile(input_filepath).map(lambda line: json.loads(line))

    # Default Partitioning

    default_start = time.time()

    top10_business_default_RDD = (reviews.map(lambda x: (x["business_id"], 1))
                                  .reduceByKey(lambda a, b: a + b)
                                  .sortBy(lambda x: (-x[1], x[0])))
    
    
    top10_business_default_RDD.take(10)
    default_end = time.time()

    default_n_partitions = top10_business_default_RDD.getNumPartitions()
    default_n_items = top10_business_default_RDD.glom().map(len).collect()

    default_duration = default_end - default_start

    # Custom Partitioning

    custom_start = time.time()

    reviews_custom = reviews.map(lambda x: (x["business_id"], 1)).partitionBy(n_partition, custom_partitioner)

    top10_business_custom_RDD = (reviews_custom
                                .reduceByKey(lambda a, b: a + b)
                                .sortBy(lambda x: (-x[1], x[0])))


    top10_business_custom_RDD.take(10)

    custom_end = time.time()

    custom_n_partitions = top10_business_custom_RDD.getNumPartitions()
    custom_n_items = top10_business_custom_RDD.glom().map(len).collect()

    custom_duration = custom_end - custom_start


    # Write results to JSON
    results = {
        "default": {
            "n_partition": default_n_partitions,
            "n_items": default_n_items,
            "exe_time": default_duration
        },
        "customized": {
            "n_partition": custom_n_partitions,
            "n_items": custom_n_items,
            "exe_time": custom_duration
        }
    }

    with open(output_filepath, 'w') as outfile:
        json.dump(results, outfile)

    sc.stop()

if __name__ == "__main__":
    main()

File: Assignments/Assignment1/task2.scala
import org.apache.spark.sql.{SparkSession, Row}
import org.apache.spark.rdd.RDD
import org.apache.spark.Partitioner

object task2 {

  class CustomPartitioner(partitions: Int) extends Partitioner {
    def numPartitions: Int = partitions

    def getPartition(key: Any): Int = {
      val k = key.asInstanceOf[String]
      k.hashCode % numPartitions
    }
  }

  def main(args: Array[String]) {
    if (args.length != 3) {
      println("Usage: spark-submit --class task2 hw1.jar <review_filepath> <output_filepath> <n_partition>")
      System.exit(1)
    }

    val reviewFilepath = args(0)
    val outputFilepath = args(1)
    val nPartition = args(2).toInt

    val spark = SparkSession.builder.appName("Assignment1_Task2").getOrCreate()

    import spark.implicits._

    val reviews = spark.read.json(reviewFilepath)
      .map(row => row.getAs[Map[String, String]]("business_id"))
      .rdd

    // Default Partitioning
    val defaultStart = System.currentTimeMillis

    val top10_business_default_RDD = reviews.map(x => (x("business_id"), 1))
      .reduceByKey(_ + _)
      .sortBy(x => (-x._2, x._1))

    top10_business_default_RDD.take(10)

    val defaultEnd = System.currentTimeMillis

    val defaultNPartitions = top10_business_default_RDD.getNumPartitions
    val defaultNItems = top10_business_default_RDD.glom().map(_.length).collect()

    val defaultDuration = defaultEnd - defaultStart

    // Custom Partitioning
    val customStart = System.currentTimeMillis

    val reviewsCustom = reviews.map(x => (x("business_id"), 1))
      .partitionBy(new CustomPartitioner(nPartition))

    val top10_business_custom_RDD = reviewsCustom
      .reduceByKey(_ + _)
      .sortBy(x => (-x._2, x._1))

    top10_business_custom_RDD.take(10)

    val customEnd = System.currentTimeMillis

    val customNPartitions = top10_business_custom_RDD.getNumPartitions
    val customNItems = top10_business_custom_RDD.glom().map(_.length).collect()

    val customDuration = customEnd - customStart

    val results = Map(
      "default" -> Map(
        "n_partition" -> defaultNPartitions,
        "n_items" -> defaultNItems.toList,
        "exe_time" -> defaultDuration.toDouble
      ),
      "customized" -> Map(
        "n_partition" -> customNPartitions,
        "n_items" -> customNItems.toList,
        "exe_time" -> customDuration.toDouble
      )
    )

    val resultsDF = Seq(results).toDF()
    resultsDF.write.json(outputFilepath)

    spark.stop()
  }
}
